### ISSUE TITLE
Metadata Schema version 2.0.0 (comments due by July 31st COB)

### ISSUE THREAD

**Issue by okamanda** [GSA/code-gov-web#280]
All,

Please take look at the following update schema, [diff run on the sample code file](https://github.com/presidential-innovation-fellows/code-gov-web/blob/master/src/assets/v200schemadiffv2.diff), and the newly [revised sample code.json](https://github.com/presidential-innovation-fellows/code-gov-web/blob/master/src/assets/draft_v200_sample_code.json). Your comments on the [schema, as proposed](https://github.com/presidential-innovation-fellows/code-gov-web/blob/master/src/assets/draftcodegovschema.json), are due on July 31st at close of business.

## Summary of Changes

- Version Number - In keeping with semantic versioning, this should increase from 1.0.2 to 2.0.0

- projects -> releases. See the comment posed by @ctubsii #250 . He teases out a distinction between projects and releases, which might make it easy to resolve some of the ambiguity about what code should be tracked and released on code.gov.
 
- licenses. The licenses field should actually be an array of license objects which holds both the license name and the URL to the license itself.

- permissions. - some have pointed out that _openSourceProject_ & _governmentWideReuseProject_  are  confusing because they allow for the possibility that a release could be open source and not available for governmentWideReuse if openSourceProject was set to 1 and governmentWideReuseProject was set to 0. Even changing these fields to type boolean rather than number doesn't resolve the ambiguities. Instead, we're proposing a _permissions_ object with a closed list of possible values based on the usageType:

- - "open" for open source

- - "governmentWideReuse" for Government Wide Reuse

- - "exemptByLaw" for the following exemption: "The sharing of the source code is restricted by law or regulation, including—but not limited to—patent or intellectual property law, the Export Asset Regulations, the International Traffic in Arms Regulation, and the Federal laws and regulations governing classified information."

- - "exemptByNationalSecurity" for the following exemption: " The sharing of the source code would create an identifiable risk to the detriment of national security, confidentiality of Government information, or individual privacy."

- - "exemptByAgencySystem" for the following exemption: "The sharing of the source code would create an identifiable risk to the stability, security, or integrity of the agency’s systems or personnel."

- - "exemptByAgencyMission" for the following exemption: "The sharing of the source code would create an identifiable risk to agency mission, programs, or operations."

- - "exemptByCIO" for the following exemption: "The CIO believes it is in the national interest to exempt sharing the source code."

- - repositoryURL & homepageURL - renaming the repository and homepage fields to be syntactically correct.

- - laborHours. This is the new field to measure the value of the code; how many hours did it take to build this release? This will replace the proposed costSavings field.

- - measurementType. This is a new field where agencies can choose from a list of how they measured the 20%. There is a list of options from which to choose and an 'other' option, after which you can describe the method used to count 20%. 

- - - linesOfcode: "Source lines of code"
- - - modules: "Number of self-contained modules"
- - - cost: "Cost"
- - - releases: "Number of software releases"
- - -systems: "System certification and accreditation boundaries"
- - - date: This replaces the 'updated' object, and swaps out _lastCommit_ for _created_.

## (New) metadata schema 2.0.0
```
{
  "$schema":"http://json-schema.org/draft-04/schema#",
  "title":"Code.gov Inventory",
  "description":"A federal source code catalog",
  "type":"object",
  "properties":{
    "version":{
      "type":"string",
      "description": "The version of the metadata schema in use. For example '1.0.2'."
    },
    "agency":{
      "type":"string",
      "description": "The agency acronym for Clinger Cohen Act agency, as defined by the United States Government Manual."
    },
    "releases":{
      "type":"object",
      "description": "An object containing each versioned source code project made available",
      "properties":{
        "name":{
          "type":"string",
          "description": "The name of the release."
        },
        "organization":{
          "type":"string",
          "description": "The organization within the agency that the releases listed belong to. For example, '18F' or 'Navy'."
        },
        "description":{
          "type":"string",
          "description": "A one- or two-sentence description of the release."
        },
        "licenses":{
          "type":"array",
          "items":{
            "type":"object",
            "properties":{
            "URL":{
              "type":"uri",
              "description": "The URL of the project license, if available. null should be used if not."
              },
            "name":{
              "type":"string",
            "description": "an abbreviation for the name of the type license. For example, 'CC0' or 'MIT'."
              }
          }
        }
        },
        "permissions":{"type":"object",
                      "properties":{
                        "usageType":{"type":"string",
                                     "description": "A list of enumerated values which describes the usage or license permissions for the release.",
                                     "enum": ["open","governmentWideReuse","exemptByLaw", "exemptByNationalSecurity", "exemptByAgencySystem","exemptByAgencyMission", "exemptByCIO"
]},
                        "exemptionText":{"type":[null,"string"],
                                        "description": "If an exemption is listed in the 'usageType' field, this field should include a one- or two- sentence justification for any exemption used."}
                     }
                      }
        ,
        
        "tags":{
          "type":"array",
          "items":{
            "type":"string",
            "description": "An array of keywords describing the release."
          }
        },
        "contact":{
          "type":"object",
          "properties":{
            "email":{
              "type":"string",
              "description": "The email address for the point of contact for the release."
            },
            "name":{
              "type":"string",
               "description": "The name of the point of contact for the release."
            },
            "URL":{
              "type":"uri",
               "description": "The URL to a website that can be used reach the point of contact. For example, 'http://twitter.com/codeDotGov'"
            },
            "phone":{
              "type":"number",
               "description": "A phone number for the point of contact for the release."
            }
          }
        },
        "status":{
          "type":"string",
           "description": "The development status of the release.",
          "enum": ["Ideation","Alpha","Beta", "Production", "Archival"]
        },
        "vcs":{
          "type":"string",
           "description": "A lowercase string with the name of the version control system in use on the project."
        },
        "repositoryURL":{
          "type":"uri",
            "description": "The URL of the public project repository for open source repositories. This field is not required for repositories that are only available as government-wide reuse or are closed, pursuant to one of the exemptions."
          
        },
        "homepageURL":{
          "type":"uri",
          "description": "The URL of the public release homepage."
        },
        "downloadURL":{
          "type":"uri",
          "description": "The URL where a distribution of the release can be found."
        },
        "languages":{
          "type":"array",
          "description": " An array of strings with the names of the programming languages in use on the release.",
          
          "items":{
            "type":"string"
          }
        },
        "laborHours":{
          "type":"number",
          "description": "The total number of labor hours spent on the development of this release."
          
        },
        "measurementType":{
          "type":"object",
          "properties":{
            "method":{
              "type":"string",
              "enum": ["linesOfcode","modules","cost", "releases", "systems", "other"],
              "description": "An enumerated list of different methods for measuring the open source requurement."
            },
            "ifOther":{
              "type": "string",
              "description":"A one- or two- sentence description of the measurement type used, if 'other' is selected as the value of 'method' field."
            }
          }
          
        },
        "partners":{
          "type":"array",
          "description":"An array of objects including an acronym for each agency partnering on the release and the contact email at such agency.",
          "items": {
            "type": "object",
              "properties":{
            "name":{
              "type":"string",
              "description":"The acronym describing the partner agency."
            },
            "email":{
              "type":"string",
              "description":"The email address for the point of contact at the partner agency."
            }
          }
            
          }
        
        },
        
        "date":{
          "type":"object",
          "description":"A date object describing the release.",
          "properties":{
            "created":{
              "type":"date",
              "description": "The date the release was created, in YYYY-MM-DD or ISO 8601 format."
              
            },
            "lastModified":{
              "type":"date",
              "description": "The date the release was modified, in YYYY-MM-DD or ISO 8601 format."
            },
            "metadataLastUpdated":{
              "type":"date",
              "description": "The date the metadata of the release was last updated, in YYYY-MM-DD or ISO 8601 format."
            }
          }
        }
      },
      "required":[
        "name",
        "permissions",
        "repositoryURL",
        "description",
        "licenseURL",
        "laborHours",
        "measurementType",
        "tags",
        "contact",
        "date"
        
      ]
    }
  },
  "required":[
    "version",
    "agency",
    "releases"
  ]
}

```
## Diff against version 1.0.2

```diff
--- sample_code.json	2017-07-24 15:55:04.000000000 -0400
+++ draft_v200_sample_code.json	2017-07-25 17:36:00.000000000 -0400
@@ -1,49 +1,58 @@
 {
-    "version":"1.0.1",
+    "version":"2.0.0",
     "agency": "DOABC",
-    "projects": [
+    "releases": [
         {
             "name": "mygov",
             "organization": "XYZ Department",
             "description": "A Platform for Connecting People and Government",
-            "license": "https://path.to/license",
+            "licenses": [{
+              "URL": "https://path.to/license",
+              "name": "CC0"
+          
+            }],
-            "openSourceProject": 1,
-            "governmentWideReuseProject": 0,
-
+            "permissions":{
+              "usageType":"open",
+              "exemptionText": null
+            },
+          
+          
             "tags": [
               "platform",
               "government",
               "connecting",
               "people"
             ],
             "contact": {
-              "email": "project@agency.gov",
+              "email": "release@agency.gov",
               "name": "Project Coordinator Name",
               "URL": "https://twitter.com/projectname",
               "phone": "2025551313"
             },
             "status": "Alpha",
             "vcs": "git",
-            "repository": "https://github.com/presidential-innovation-fellows/mygov",
+            "repositoryURL": "https://github.com/presidential-innovation-fellows/mygov",
-            "homepage": "https://agency.gov/project-homepage",
+            "homepageURL": "https://agency.gov/project-homepage",
             "downloadURL": "https://agency.gov/project/dist.tar.gz",
             "languages": [
               "java",
               "python"
             ],
+            "laborHours": "1000",
+            "measurementType": "linesOfcode",
             "partners": [
                 {
                     "name": "DOXYZ",
-                    "email": "project@doxyz.gov"
+                    "email": "releaseContact1@doxyz.gov"
                 }
             ],
-            "exemption": null,
-            "exemptionText": "No exemption requested",
-            "updated": {
-                "lastCommit": "2016-04-30",
-                "metadataLastUpdated": "2016-04-13",
-                "lastModified": "2016-04-12"
+            
+            "date": {
+                "created": "2016-04-12",
+                "lastModified": "2016-04-12",
+                "metadataLastUpdated": "2016-04-13"
+                
             }
         }
     ]
 }
```

## New Sample Code.json (v.2.0.0)
```
{
    "version":"2.0.0",
    "agency": "DOABC",
    "releases": [
        {
            "name": "mygov",
            "organization": "XYZ Department",
            "description": "A Platform for Connecting People and Government",
            "licenses": [{
              "URL": "https://path.to/license",
              "name": "CC0"
          
            }],
            "permissions":{
              "usageType":"open",
              "exemptionText": null
            },
          
          
            "tags": [
              "platform",
              "government",
              "connecting",
              "people"
            ],
            "contact": {
              "email": "release@agency.gov",
              "name": "Project Coordinator Name",
              "URL": "https://twitter.com/projectname",
              "phone": "2025551313"
            },
            "status": "Alpha",
            "vcs": "git",
            "repositoryURL": "https://github.com/presidential-innovation-fellows/mygov",
            "homepageURL": "https://agency.gov/project-homepage",
            "downloadURL": "https://agency.gov/project/dist.tar.gz",
            "languages": [
              "java",
              "python"
            ],
            "laborHours": "1000",
            "measurementType": "linesOfcode",
            "partners": [
                {
                    "name": "DOXYZ",
                    "email": "releaseContact1@doxyz.gov"
                }
            ],
            
            "date": {
                "created": "2016-04-12",
                "lastModified": "2016-04-12",
                "metadataLastUpdated": "2016-04-13"
                
            }
        }
    ]
}
```

**Comment by ckaran** [comment]
A minor suggestion for the description of the version field of the JSON schema.  Instead of 

> "The version of the metadata schema in use. For example '1.0.2'."

how about

> "The version of the metadata schema in use. Obeys Sematic Versioning 2.0.0 rules as defined at http://semver.org/"

---

Separately from above, does it make sense for `measurementType` to be within a release?  It implies that in release `1.0.0` of a project I might be using `linesOfcode`, but by release `3.0.0` I decide to switch to `releases`; how do we handle that kind of a switch?

I understand the desire for the 20% goal, but I think that measuring cost saving might be better in the long run.  Here's my proposal for how to do it:

First, add the new field `reusedSubprojects` to the schema.  It is a list of maps.  Each map has the following keys: `projectName`, `projectURL`, `hoursToAdopt`.  `projectName` and `projectURL` uniquely tell which project's code was reused, and where to find it.  `hoursToAdopt` is how long it took to incorporate the code into the current project; this is where we measure the cost savings from.

At this point, we can measure cost savings pretty much directly; look at how much time a project saved a different project in terms of effort.  For example if project `ABC` took 1000 hours to create, and if project `XYZ` took 50 hours to adopt it, then `ABC` saved `XYZ` 950 hours of effort.  On the other hand, if `ABC` was poorly written, then it won't be adopted anywhere, meaning that it cost the agency 1000 hours, and the Government didn't get any extra benefit (if it was poorly written but adopted, then the amount of time it takes to adopt the code will increase, making `ABC` contribute less to overall agency numbers).  

Here are the advantages to this method:
- Strongly encourages good coding practices as bad code is less likely to be reused.
- Encourages agencies to look to reusing good code from others; everyone likes to show that they are saving money, and this is a simple way to show it.
- Much harder to game the system; since others have to adopt your code for your numbers to look good, you can't just choose a measure that lets you meet the 20% goal without actually meeting the purpose of that goal (sharing code and increasing reuse).
- Code.gov can crunch the numbers automatically, producing reports that Congress and others can immediately understand.

The main disadvantage that I can think of is that there will always be small projects that are critical to an agency, but only to that agency.  They will end up being somewhat penalized by this method.

**Comment by iadgovuser1** [comment]
Looks like laborHours is required. Would prefer that it is optional.

**Comment by ckaran** [comment]
@iadgovuser1 `laborHours` makes the story of why the Government would be interested in Open Source at all.  Open Sourcing material takes some effort (which means money), so unless someone can document that there is a net gain to the Government to Open Source its projects, there won't be much of an argument for it.

**Comment by IanLee1521** [comment]
(Copying from #273):  

[A] minor clarification question is: are the ``license.name`` values expected to be selected from a list? If not I would worry that you might get difference in spellings / expansions, e.g. "GPL" vs "GPL2" vs "GNU Public License" vs ...

**Comment by IanLee1521** [comment]
RE: the discussion around ``laborHours``, I wonder if a better way to estimate this (as that is all it is ever likely to be) would be to calculate based on certain statistics from the repo.

Here is a made up example with data I pulled from https://github.com/dhs-ncats/pshtt/commit/00ff246f40acbea185d478d838c7fcd6652b9aa8 (294 commits with +6944, -5163 line changes [created with this gist](https://gist.github.com/IanLee1521/fead3d955930a36ac9d53d6e1fbd4722)) using a formula I'm just making up:

```python
sum(
  292 commits * 2 minutes per commit,
  (6944 lines added + 5163 lines deleted) * 2 minutes,
  49 pull requests * 20 minutes,
  40 issues * 15 minutes,
) 
= (292 * 2) + ((6944 + 5163) * 2) + (49 * 20) + (40 * 15) = 26378 minutes 
~= 440 hours
```

Otherwise, I think it's going to very quickly get intractable for how to keep that number up to date. especially since arguably the whole point is to get people to contribute to the code, therefore the ``laborHours`` will always be a moving and inaccurate target.

Additionally, it looks like there is already some research in this area:

- https://pdfs.semanticscholar.org/3db5/64c7c379f11f308075f4c8704809ad39bc77.pdf
- https://gsyc.urjc.es/~grex/repro/2014-msr-effort/msr14-robles-estimating-effort.pdf

I also wouldn't be surprised if GitHub (cc/ @jbjonesjr / @benbalter) had a decent handle on this already...

**Comment by IanLee1521** [comment]
Also, as a single data point, putting that comment together apparently took me about 25 minutes worth of work (with a bit of getting distracted with other things).

**Comment by david-a-wheeler** [comment]
I don't think you're suggesting it, but just to be clear, please don't use made-up equations with made-up coefficients. All models are wrong, but some models are useful. But models can only be useful if they are calibrated to reality. You need to use data from reality to calibrate the model you're using, whatever it is. SloCCount for example is an implementation of Barry Boehm's COCOMO model, which is calibrated from real projects, and its calibration is published.  Good luck whatever you choose to do.

**Comment by okamanda** [comment]
Good catches, all!

laborHours, costSavings, and all other alternatives are at best approximations of value.

@ianlee1521, an enumerated list for license.name could work as long as we allow for the possibility of custom or non-standard licenses.

@ckaran, agree on your note about versioning; and measurementType should move out of the releases object

**Comment by iadgovuser1** [comment]
@ckaran  

"makes the story of why the Government would be interested in Open Source at all."

The fact the code is tax payer funded should be sufficient enough motivation to make the code open source. laborHours and costSavings will likely be highly inaccurate and/or arbitrary. They can also be used against you (e.g. it only saves us X hours/Y dollars so why are we open sourcing it or we aren't going to use that).

**Comment by ckaran** [comment]
@iadgovuser1 I agree that since the code was taxpayer funded it should be open sourced, but I also know what it takes to sell a story to upper management.  In this case 'upper management' is Congress, which tends to be unhappy if they see wasted money.  Assuming you already have a good case for why you had to custom develop the code, if you can show how much money you saved by reusing the code that others have provided, then you've got a good story to give to Congress and anyone else that asks.

As for being inaccurate, it doesn't have to be.  There are a very large number of time clock programs on the internet that will allow a developer to track his or her time spent on a given project (see https://www.google.com/search?q=open+source+time+clock+program&btnG=Search&hl=en&gbv=1 http://blog.capterra.com/the-top-8-free-and-open-source-time-clock-software-systems/).  Choose your favorite, and your development team can track what they're doing.  This can also be used to estimate how much time external contributors have saved you.  E.g., it took 83 hours of work/1000 lines of code for our internal developers, external contributors gave us 10,000 lines of code, therefore we saved 830 hours of work. 

Finally, how can it be used against you?  If you've already made the case for why you have to develop the program, then no-one can use the hours of time spent creating the program against you.  And given the level of effort it will take to get through the administrative hurdles to Open Source something, I suspect that small projects which would only save a few hours of effort won't be Open Sourced, while large projects will be, and since they will have taken the most time to develop, they may save others large amounts of time by being reused.

In short, `laborHours`, while an imperfect measure, is likely to be one of the best measures we can use.

**Comment by iadgovuser1** [comment]
@ckaran 

I'm not completely against it despite its good and bad use cases. I'm against making it a required element of the schema.

**Comment by ckaran** [comment]
@iadgovuser1 

Got it, I misunderstood.  What would you recommend projects use as a metric?  Or would you prefer that projects not use any metric at all?

**Comment by apyle** [comment]
Some thoughts.

Permissions - does this mean that any source code where an agency does have the data rights to the code falls under "exemptByLaw"? It would be good to add an "underReview" value for those instances where an agency is still trying to determine if they have the rights to share code with other agencies or open source it.

Labor Hours - since this represents the number of hours to produce this release, does that mean that the number gets reset after every release? For a team practicing continuous delivery, this number will be really small. Even when a team is creating a release every 6 months a reset will under represent the true number of hours a release represents.

It's important to remember that there are a number of tasks that go into the creation of a release besides writing code. We need to account for the time and effort for usability, security testing, contractual validation & verification, Assessment & Authorization (A&A) just to name a few items besides the time developers devote to the project.

Measurement Type - Since there will be one and only one way to measure code across an agency, this field should be at the top level and not within the release object. It will have the same value for every entry in `release`.

**Comment by klever4ever** [comment]
At SSA, we are requiring agency code assets to have a metadata file that expands on the data required to generate the code.json file. The file, codeMetadata.json, will be required at the root of each repo (similar to a readme.md).  We harvest the codeMetadata.json from the various repos in the agency to create our code inventory of code assets. We have a process that outputs our code assets from our codeMetadata.json files into the schema required by OMB. I figure this might provide some ideas for v2.0 of the code.json schema.

Here is our first version of our codeMetadata.json schema shown in md format (our codeMetadataSchema.json file available on request):

# codeMetadata.json Schema v1.0

The schema defines the following properties:

## `publiclySharedMetadata` (object, required)

The information in this section may be shared with public.

Properties of the `publiclySharedMetadata` object:

### `conformsToSchemaUrl` (string)

The URL of the schema that this metadata file conforms to.

### `codeAssetGUID` (string, required)

A globally unique identifier assigned to this code asset that will be used to identify it from this point forward, therefore it shouldnt change. If you are updating an existing codeMetadata.json file, use the GUID from that file.

### `codeAssetGroupGUID` (string, required)

A globally unique identifier assigned to this code asset group that will be used to identify it from this point forward, therefore it shouldnt change. If you are updating an existing codeMetadata.json file, use the GUID from that file. If you want to relate a few code assets together, then give them the same group GUID. For example, if your application has a backend web services module and a front end web gui module, each module should have it's own codeMetadata.json file, but they can share the same Code Asset Group GUID. 

### `codeAssetTitle` (string, required)

The name of the code asset.

### `codeAssetDescription` (string, required)

The description of the code asset.

### `languageNames` (array, required)

The language(s) used to develop this code asset.

The object is an array with all elements of the type `string`.

### `isDeveloperSandbox` (boolean)

Is developer sandbox? For example, is this code asset just some exploratory/learning code?

### `isProofOfConcept` (boolean)

Is Proof of concept? For example, is this an experiment or part of a pilot?

### `developmentStatus` (string, enum, required)

The development status of the project.

This element must be one of the following enum values:

* `Active`
* `Archival`
* `Retired`

### `partneringAgencies` (array)

An array of objects including an acronym for each agency partnering on the project and the contact email at such agency.

The object is an array with all elements of the type `object`.

The array object has the following properties:

#### `name` (string)

The acronym describing the partner agency.

#### `email` (string)

The email address for the point of contact at the partner agency.

### `publiclyHostedInstanceURLs` (array)

If this code asset compiles into a publicly hosted website or page, then it should be listed here. For example, if this code assest is representing the code behind ssa.gov/data, then ssa.gov/data should be listed as a publicly hosted url.

The object is an array with all elements of the type `string`.

### `publicRepository` (object)

Properties of the `publicRepository` object:

#### `projectHomepageURL` (string)

The URL of the public code asset homepage.

#### `downloadURL` (string)

The URL where a distribution of the code asset can be found.

#### `versionControlSystemCode` (string)

The name of the Version Control System (VCS) that is used. For example: BitBucket, GitHub, Sourceforge, etc.

#### `repositoryUrl` (string)

The URL of the public code asset repository.

#### `sourceCodeLastModifiedTimestamp` (string)

The date in YYYY-MM-DD format that the source code or package was last updated. Estimated dates are okay.

### `tags` (array, required)

A list of string alphanumeric keywords that identify the code asset. Include type of application, frameworks, keywords and functionality.

The object is an array with all elements of the type `string`.

### `concernsWithPublishingMetadata` (boolean, required)

Just to clarify, if you beleive the existence of this code asset should be hidden from the public eye, select yes.

Default: `false`

### `concernsComments` (string)

If you have concerns, please eneter them below.

## `internalMetadata` (object, required)

The information in this section is for internal use and will not be released to public.

Properties of the `internalMetadata` object:

### `organizationName` (string, required)

Enter the organization name to which the code asset belongs.

### `metadataSteward` (employee, required)

The person responsible for filling out and maintaining this file.

### `metadataUpdatedDate` (string, required)

The date this codeMetadata.json file was last updated.

### `relatedIdentifiers` (array)

Add whatever identifiers you have that are related to this code asset.

The object is an array with all elements of the type `object`.

The array object has the following properties:

#### `identifierTypeCode` (string, enum)

Identifier Code.

This element must be one of the following enum values:

* `OPN`
* `RAS`
* `IMT`
* `TRX`
* `CMS`
* `CCB`
* `SRC`
* `CAPRS`
* `EMATS`

#### `value` (string)

Identifier Value.

### `dependentOnDataStore` (boolean)

Does the source code depend on a Data Store? (Does this code need a database or similar to work properly?)

### `readsPII` (boolean)

Does the source code read Personally identifiable information (PII)?

### `writesPII` (boolean)

Does the source code write Personally identifiable information (PII)?

### `readsPHI` (boolean)

Does the source code read Protected health information (PHI) ?

### `writesPHI` (boolean)

Does the source code write Protected health information (PHI) ?

### `readsFTI` (boolean)

Does the source code read Federal Tax Information (FTI)?

### `writesFTI` (boolean)

Does the source code write Federal Tax Information (FTI)?

### `embeddedPasswords` (boolean)

Does the source code have embedded Passwords?

### `embeddedEncryptionHash` (boolean)

Does the source code have embedded Encryption Hash?

### `embeddedServerConfiguration` (boolean)

Does the source code have embedded Encryption Hash?

### `relatedCOTS` (array)

If this code asset depends on or requires a COTS product or similar closed-source library, they should be listed here.

The object is an array with all elements of the type `string`.

### `solutionArchitects` (array)

Solution Architect(s) information.

The object is an array with all elements of the type `employee`.

### `developers` (array)

Developers information.

The object is an array with all elements of the type `employee`.

### `developmentStartDate` (string, required)

The date when development of this code asset started (regardless of release or version). Estimates are okay.

### `developmentEndDate` (string)

If develpment on the code asset is no longer active, the date when development stopped. Estimates are okay.

### `internalRepository` (object, required)

Source Code Repository Information.

Properties of the `internalRepository` object:

#### `versionControlSystemCode` (string, enum, required)

Version Control System Code.

This element must be one of the following enum values:

* `Bitbucket`
* `Git`
* `MKS`
* `TFS`
* `Endevor`
* `File System`

#### `repositoryUrl` (string, required)

Repository URL or location.

### `intranetProjectHomepageURL` (string)

Intranet Project Homepage URL.

### `hostedInstances` (array)

Hosted instances(Development/Validation/Production/etc.)

The object is an array with all elements of the type `object`.

The array object has the following properties:

#### `name` (string)

Name of the hosted instance environment.

#### `uri` (string)

URI/System information of the hosted instance.

---

# Sub Schemas

The schema defines the following additional types:

## `employee` (object)

Properties of the `employee` object:

### `name` (string, required)

The name of the employee.

### `emailAddress` (string, required)

The email address of the employee.

### `company` (string, required)

The company of the employee.

**Comment by klever4ever** [comment]
A few nuances
> '-    "projects": [
> '+    "releases": [

1. I'm glad we are renaming projects, I felt that naming was more budget/line item sounding. One "project" can contain many separate pieces of code. However, i feel "releases" is to specific. I think we should refer to them generically as "codeAssets". 

2. I think we should add a field after version...conformsToSchemaUrl (string) -- The URL of the schema that this metadata file conforms to.

3. Should the "measurementType" be outside of the individual "projects/release/?codeAsset?" and placed at the same level as the agency? Also, suggest a comment field for the measurementType. I also think maybe prefix w/ compliance or something?

Example of what my proposed changes would look like:
{
    "version":"2.0.0",
    "conformsToSchemaURL": "http://locationofschemainjsonformat"
    "agency": "acronymOfAgency",
    "complianceMeasurementType" : "hybrid",
    "complianceMeasurementTypeComment": "At the department of frozen yogurt, we measure"
    "codeAssets": [

    ]

**Comment by okamanda** [comment]
@klever4ever Thanks for sharing the SSA metadata schema. QQ re: `conformsToSchemaURL`. If the release/project/repo version is already defined as 2.0.0 (or 1.0.1) then why would we need another external link to that schema file?

**Comment by klever4ever** [comment]
I don't know for sure if it's a best practice or not. I got the idea from seeing what was done for Project Open Data. @philipashlock can probably chime in and give input.

I'm noticing that I mixed up two of the related fields in my recommendation:
https://project-open-data.cio.gov/v1.1/schema/#conformsTo
https://project-open-data.cio.gov/v1.1/schema/#describedBy

I meant to recommend/was thinking of describedBySchemaURL. Both are useful to have IMO. In any event, the concept I was thinking is the same: reference the json schema file itself inside of the code.json 

## Why?:
1. Allowing agencies to extend the code.json schema itself and document accordingly. IIRC, back when we were at v1.0 ish times @AlvandSalehi mentioned that we can extend the code.json schema as needed if we want include properties that aren't in the 'official ' schema. @AlvandSalehi , correct me if i'm misremembering. 
2.  The consumers of agency's code.json (OMB for code.gov) can write automated validators to validate against the schema json referenced in describedBySchemaURL. (...r u thinking "can't we do that from the schema version?"? ... yes of course...but you'd have to hardcode your schema validation into whatever code you have validating our schemas, right? You wouldn't have to hardcode it in with this approach...and a bigger benefit IMO would be: 
3. If someone else decides to start harvesting agencies code.json files, they can utilize tools to autogenerate classes from the json schema url. examples:
http://json2csharp.com/
http://www.jsonschema2pojo.org/
https://pypi.python.org/pypi/warlock
4. general validation: http://json-schema.org/implementations.html 

As an example for 1 above, if SSA wanted to add 'id' for each project/codeasset/etc code.json . We're technically not following OMB's 2.0. We're following an extended version of your schema which we would reference accordingly in our describedBySchemaURL (brainstorming something like "https://ssa.gov/developer/code/schema/v2.0/code.schema.json"). Inside of the ssacode.schema.json would be json schema syntax that extends your schema and would reference it accordingly (see the address example: https://spacetelescope.github.io/understanding-json-schema/structuring.html) 

`"id": "US-GOV-SSA-DATA-71"`
`"name": "mygov",`             
 `"organization": "XYZ Department",`            
` "description": "A Platform for Connecting People and Government",`

### take a quick skim:
https://en.wikipedia.org/wiki/JSON#Schema_and_metadata

Doesn't bother me one way or another if we go w/ what I'm proposing. I'm sure I'm recommending a bazooka to kill a moth.

# 🥊 kapow 🦋

**Comment by iadgovuser1** [comment]
@ckaran I'd go with whatever is easiest to automate so in this case probably the releases metric.

**Comment by ckaran** [comment]
@apyle brought up a really good point about how `laborHours` could be hard to measure with continuous releases.  My suggestion is that it be the total number of hours expended up to that point in time.  **However** for this to work, each project would have to maintain its own `code.json` file within its repository.  As work progresses, the number of labor hours can be increased.  If another project comes along and uses release 2.0 of a project, then the client will use the labor hours up to that point in its own metrics.  This way, if the client never uses a more advanced version of the library, it isn't skewing the statistics.  This will work even with continuous release models, and is actually relatively easy to automate; I just walk through all the projects I'm using, scanning for their `code.json` files, and build my own `code.json` file from them.  This will also handle the continuous release model properly.

The reason that I prefer this over the method suggested by @iadgovuser1 is that the number of releases doesn't capture the level of effort required to bring a new release out.  As an example, I'm currently using Firefox release 55.  Given their release schedule, I can expect another release sometime in the next few weeks.  On the other hand, the latest Linux kernel release is in the 4.x range.  How do you compare the level of effort to go from the 3.x to the 4.x Linux kernel series vs. the level of effort to go from Firefox 54 to 55?

Finally, `laborHours` is inflation proof; regardless of what inflation does in the future, 1 hour is 1 hour.  People won't emotionally discount it as being cheap simply because they can't do inflation adjusted numbers in their heads quickly.

**Comment by DanielJDufour** [comment]
Thanks for all the great ideas.  I'm closing this now because of inactivity, but happy to continue the conversation in another ticket as issues come up.
